### DISCUSSION TITLE
Console App to get workflow definition from server, run it (on the client) and update server of the run (new instance)

### DISCUSSION THREAD

**Discussion by alrodi** [elsa-workflows/elsa-core#4452] [Q&A]
I would like to keep the server to create, manage and report the workflows but not run then. I want several client console apps, which are triggered by other processes to read the definition for an specific workflow, execute the workflow and report the status of each activity to the server in order to have a central place to manage, see the workflows executions. Is this possible? I cannot find REST API methods to report the workflow as executed or sync the execution steps... Am I missing something? I do not want my execution to happen triggered or ran from the server. Thanks in advance for any help.

**Comment by sfmskywalker** [comment]
Hello @alrodi 

The system allows for this, but just to be clear: you want to essentially load the workflow definition into your console app, run it from the console app, and then save the workflow instance back into the server. Is that right?

If yes, then the only thing I think we're missing is an API endpoint to receive the workflow instance. Right now, the WorkflowInstance API resource only supports the following operations:

- Bulk Cancel
- Bulk Delete
- Delete
- Get
- Journal
- List

So, we would need to add:

- Post

I think your scenario makes a lot of sense! I'll open a new issue with this change request.

Thanks for the question!

**Reply by sfmskywalker** [reply]
Issue created [here](https://github.com/elsa-workflows/elsa-core/issues/4453).

**Reply by alrodi** [reply]
Yes, exactly! thanks!. Does the workflow instance keeps a log of all the executed activities? My goal is to run these cashflows and at the end provide a link to see the workflow instance execution in the form of a multilevel collapsible list (like a github build pipeline), where you can drill down from top level activities to sub activities. I understand that in your current functionality I can see the workflow instance state/execution in a graphic form, I am not sure if these activity states are persisted and easily readable to build a control like that. In case it is not, is there a logging middleware where I can intersect the pre/post of the activity which includes the parent activities, status, activity name, description, variables, etc? Also, I imagine I can segregate security in such a way that the consuming console app would only have permissions to read the definition and post instances but nothing else (i.e. delete a workflow). Thanks in advance.

**Reply by sfmskywalker** [reply]
Yes on all accounts 😄 

- Yes, there are two tables that log workflow execution: ActivityExecutionRecords and WorkflowExecutionLogRecords. The former collects state about when an activity started and when it completed inside of a the same record, while the latter represents individual events as records (e.g. Started, Completed, Faulted, and custom events).
- Yes, you can write a custom middleware component to capture data as an activity executes. In fact, that's precisely what Elsa 3 leverages for capturing activity execution records and workflow execution log records.
- Yes, each API endpoint configures required permissions. These permissions can be configured per user role and per application, and typically translate into JWT bearer tokens containing the "permissions" claim and the set of effective permissions. The authentication & authorization mechanism leverages ASP.NET Core services, which means you can use whatever ASP.NET Core supports - you're not limited to OpenID Connect / JWT, etc.
